web_app_drawer/static/src/js/web_app_drawer.js

```diff
@@ -0,0 +1,183 @@
+        currentPos: [0, 0],
+        
+        init: function() {
+            this.$el = $('.drawer');
+            this.$el.drawer();
+            this.directionCodes = {
+                37: this.LEFT,  // Left arrow
+                39: this.RIGHT,  // Right arrow
+                38: this.UP,  // Up arrow
+                33: this.UP,  // page up
+                40: this.DOWN,  // Down arrow
+                34: this.DOWN,  // page down
+            };
+            var $clickZones = $('.openerp_webclient_container, ' +
+                                'a.oe_menu_leaf, ' +
+                                'a.oe_menu_toggler'
+                                );
+            $clickZones.click($.proxy(this.handleClickZones, this));
+            this.$el.one('drawer.opened', $.proxy(this.onDrawerOpen, this));
+            // Core bus keypress doesn't seem to send arrow keys
+            core.bus.on('keydown', this, this.handleNavKeys);
+        },
+        
+        // It provides handlers to hide drawer when "unfocused"
+        handleClickZones: function() {
+            this.$el.drawer('close');
+            $('.oe_secondary_menus_container')
+                .parent()
+                .collapse('hide');
+        },
+        
+        // It provides keyboard shortcuts for app drawer nav
+        handleNavKeys: function(e) {
+            if (!this.isOpen){
+                return;
+            }
+            if (Object.keys(this.directionCodes).indexOf(e.keyCode.toString()) !== -1) {
+                var $link = this.findAdjacentAppLink(
+                    this.$selectedAppLink,
+                    this.directionCodes[e.keyCode][0],
+                    this.directionCodes[e.keyCode][1]
+                );
+                this.selectAppLink($link);
+            }
+        },
+        
+        onDrawerClose: function() {
+            this.isOpen = false;
+            this.$selectedAppLink = $(false);
+            core.bus.trigger('drawer.closed');
+            this.$el.one('drawer.opened', $.proxy(this.onDrawerOpen, this));
+        },
+        
+        // Generate x and y coords for app icons
+        onDrawerOpen: function() {
+            var self = this;
+            this.isOpen = true;
+            this.$appLinks = $('.css-app-drawer-icon-app').parent();
+            this.selectAppLink($(this.$appLinks[0]));
+            this.linkMap = {};  // Key by row, col
+            var x = 0,
+                y = -1,
+                lastTop = -9999,
+                newTop = null,
+                $lastVal = null;
+            _.each(this.$appLinks, function(val) {
+                $lastVal = $(val);
+                newTop = $lastVal.offset().top;
+                if (lastTop < newTop) {
+                    y ++;
+                    x = 0;
+                    self.linkMap[y] = {};
+                }
+                self.linkMap[y][x] = $lastVal;
+                $lastVal.data('pos-x', x).data('pos-y', y);
```

Comment:
I think this is over-engineered.

We have 2 possibilities for keyboard selection: left/right and up/down.

For L/R: just use [`prev`](http://api.jquery.com/prev/) and [`next`](http://api.jquery.com/next/) with a bit of traversing -- `.closest("li")` and so on.

For U/D, see [this example](http://stackoverflow.com/a/31643904/1468388). You can use jQuery to know how many items per row are there. Then just `.closest("ul").children(":nth-child(" + ($currentIcon.index() - iconsPerRow) + ")")` and some more traversing.

And all of this should get triggered when user presses the keyboard arrow, not when opening drawer. Otherwise imagine if I open the drawer and then resize the browser window, and then use the keyboard.